### DISCUSSION TITLE
A bidding module that enables auctions without the need for a trusted third party

### DISCUSSION THREAD

**Discussion by cryptograffiti** [btcpayserver/btcpayserver#1919] [Plugins & integrations]
Problem: Art and collectables are difficult to value.  Current options for auctions using BTC/LN involve trusting a third party.

Solution: A bidding module that utilizes a deposit to enable auctions without the need for a trusted third party.

- - - 

Sellers of an item submit the following:

- auction title 
- reserve price
- auction end time
- bitcoin address for bid deposits

Bidders of an item submit the following:

- initial bid matching or exceeding the pre-determined reserve price
- deposit amount pre-determined by seller 
- bitcoin address or LN channel for verifying the winner
- some form of contact information 

Bids are publicly displayed from highest amount to lowest amount.  Contact info is only viewable by the seller.

A countdown timer tracks the auction end time.

Bids made within 10 minutes of auction close automatically extend the auction an additional 10 minutes. 

Upon auction close, the seller contacts the winner and verifies they have won by having them provide the aforementioned BTC address or LN invoice submitted with their bids.  

The winner pays the balance (bid amount less deposit) to the bitcoin address or requests repayment via LN invoice.

Should the winner not contact the seller by a pre-determined amount of time, the seller contacts the next highest bidder until a sale is finalized.

After a sale is finalized, the seller contacts the losers of the auction to refund their bid amount via the BTC address or LN address provided.

**Comment by Zaxounette** [comment]
A few suggestions/ideas that come up in my mind:

> reserve price

A reserve price can potentially be undisclosed. Give the option to enter a reserve price but hash it and display only the hash publicly and disclose the unhashed version automatically once timer ends for transparency.

>bitcoin address for bid deposits

A non-custodial deposit system might be tricky.
Maybe a time-stamped contract ?
Or a DLC, like the one @NicolasDorier is experimenting with ? 
Could be used for the whole bid process, instead of having 2 transactions (one for deposit + one for bid finalization)

>initial bid matching or exceeding the pre-determined reserve price

In some auctions, buyer doesn't know the reserve price, so limiting them to only bid above in the UI can give that info away.

>bitcoin address or LN channel for verifying the winner

I don't see this as a reliable way of identification.
BTW, why use this if you have a contact form as proposed ? Wouldn't it be dual-infos ?
BTCPay Server works on invoices, so when generating an invoice for a bid, devs could just make input fields for contact info, like the current input field for a refund e-mail in invoices.

>Bids are publicly displayed from highest amount to lowest amount. Contact info is only viewable by the seller.

Make this an option. Some auctions can be double blinded, no publicly visible bids.

>Bids made within 10 minutes of auction close automatically extend the auction an additional 10 minutes.

_Suggestion:_ Make this a customisable feature.
Extend timer on bid TRUE/FALSE
Set entend time 1min/5mins/10mins/other

>Should the winner not contact the seller by a pre-determined amount of time, the seller contacts the next highest bidder until a sale is finalized.
>After a sale is finalized, the seller contacts the losers of the auction to refund their bid amount via the BTC address or LN address provided.

DLC/Smart Contract could be time-locked and release funds after XYYZ days/blocks/chicken have had eggs.


All in all, to make the dev's hair turn grey, you could also ask for different auction types implemented directly: 
* Standard
* Dutch
* Reversed Dutch
* Etc ...

I love the idea, impatient to see other's inputs :)

**Comment by Eskyee** [comment]
This idea sounds very interesting ✨

**Comment by cryptograffiti** [comment]
Thanks for all the great suggestions.  I was thinking a more bare bones version would have a higher likelihood of getting made :)
> 
> > bitcoin address for bid deposits
> 
> A non-custodial deposit system might be tricky.

I imagined this being custodial and varying depending on what the seller deems large enough to thwart bid spammers, but small enough to overcome trust barriers.
> 
> > bitcoin address or LN channel for verifying the winner
> 
> I don't see this as a reliable way of identification.

> BTW, why use this if you have a contact form as proposed ? Wouldn't it be dual-infos ?

Yes, in case the winner does not follow through with the purchase, I thought the contact info would be needed to then contact the next highest bidder.

This is my first time using Github so please excuse any formatting mishaps and thanks for looking!

**Comment by dstrukt** [comment]
I thought this idea was really interesting, so I spent an hour mocking out a rough wireframe with all of the ideas presented above.

Big questions definitely around:
- User Auth / Verification
- Shipping / Returns / the usual P2P hand-off issue(s)
- The third-party or time-locked custodian of bids / funds ... I don't think someone should have control of releasing these funds, rather the money is returned when the auction ends or after X period of time from the initial bid

Obviously quite a lot still to do, but I thought this would get us started to have some more concrete discussions :)

Happy to hear thoughts.

![Auction](https://user-images.githubusercontent.com/6250771/94912741-fa6ecd80-045c-11eb-8a44-848522f9b06c.png)

![Auction - Place Bid](https://user-images.githubusercontent.com/6250771/94912738-f9d63700-045c-11eb-9aa8-25f7696882f0.png)

![Apps - Auction](https://user-images.githubusercontent.com/6250771/94912734-f80c7380-045c-11eb-9e19-5b374a072f19.png)

**Comment by Eskyee** [comment]
Beautiful... 😍 now take my money 💰

**Comment by Zaxounette** [comment]
> * User Auth / Verification

`LN auth` and such could be used ?
Pretty sure it's still (like LN in general) alpha/beta stages. Would limit the auction to tech-savvy people though.

> * Shipping / Returns / the usual P2P hand-off issue(s)

That issue will likely have to be handled by the auction creator on a case by case.
Automation could be included with an e-mail automatically sent to the winner asking for shipping details etc ...
Potentially give the creator the option to use that automation, or by default let them contact the winner directly with the e-mail provided at bid.

> * The third-party or time-locked custodian of bids / funds ... I don't think someone should have control of releasing these funds, rather the money is returned when the auction ends or after X period of time from the initial bid

This issue could become a use-case of Hodl-invoices (https://lightningwiki.net/index.php/HODL_Invoice + https://wiki.ion.radar.tech/tech/research/hodl-invoice ). 
Funds stay locked until auction finishes, and unlocked and spendable again by owners of those funds that didn't win auction. 
_Just an idea._

> Obviously quite a lot still to do, but I thought this would get us started to have some more concrete discussions :)

I really like the UI/UX you presented here ! <3


_Also, food for though:_ Some auctions can go very quickly, especially in the last minutes/secondes => Make the app increment bids at the time the `place bid` is clicked, not at time of payment and have it update dynamically to avoid double/triple/quadrupal same priced bids.

**Comment by NicolasDorier** [comment]
I don't see anyway of doing that without trusted third party.
The only thing a DLC can do (and DLC rely on a third party, the oracle) is more akin to crowdfunding:

The oracle would define a DLC with two outcome "Entirely funded" or "Not funded".
Backers would engage in DLCs with the Funded party.
When it reach the goal, the oracle would attest the outcome Entirely funded, effectively giving the money to the funder.

But even this, is not without flaws: A malicious Funded party could fake backing by engaging a DLC to himself, so the oracle get tricked into thinking it is entirely funded.

**Comment by Eskyee** [comment]
If we could solve this third party problem !! Every artist would have a real tool under their belt... plus sites like https://www.artsy.net and others would take notice of the bitcoin network 😍

**Comment by cryptograffiti** [comment]
> I don't see anyway of doing that without trusted third party.

Agreed, some level of trust between buyer and seller would be needed and I think that would be okay in many scenarios.

**Comment by Eskyee** [comment]
I would personally trust the independent artist all day long.. they have work hard for what they achieved... and should be fully rewarded. Achieving this with BTCPAY software is the end goal 🥅 and would be a dream 😍 come true for most.

**Comment by cryptograffiti** [comment]
> I thought this idea was really interesting, so I spent an hour mocking out a rough wireframe with all of the ideas presented above.
> 
This looks really cool!!  Thanks for doing it

**Comment by dstrukt** [comment]
Thanks - no problem. 

Even if we might not be in a place to implement this / the third-party trust considerations might inhibit things right now, the work is on our Figma, and can be referenced or refined when the time comes!

**Comment by cryptograffiti** [comment]
> Thanks - no problem.
> 
> Even if we might not be in a place to implement this / the third-party trust considerations might inhibit things right now, the work is on our Figma, and can be referenced or refined when the time comes!

Just posted a bounty from @nvk and I.  Hopefully that inspires :) https://chat.btcpayserver.org/btcpayserver/pl/oa1n4bp3qpddmn5b5exbirhzeo
